libbeat/common/file/rotator.go

```diff
@@ -392,70 +316,176 @@ func (r *Rotator) purgeOldSizedBackups() error {
 	return nil
 }
 
-func (r *Rotator) rotate(reason rotateReason) error {
-	if err := r.closeFile(); err != nil {
-		return errors.Wrap(err, "error file closing current file")
+func (r *Rotator) isRotationTriggered(dataLen uint) (rotateReason, time.Time) {
+	for _, t := range r.triggers {
+		reason := t.TriggerRotation(dataLen)
+		if reason != rotateReasonNoRotate {
+			return reason, time.Now()
+		}
 	}
+	return rotateReasonNoRotate, time.Time{}
+}
 
-	var err error
-	if r.intervalRotator != nil {
-		// Interval and size rotation use different filename patterns, so we use
-		// rotateByInterval if interval rotation is enabled, even if this specific
-		// rotation is triggered by size.
-		err = r.rotateByInterval(reason)
-	} else {
-		err = r.rotateBySize(reason)
+// Sync commits the current contents of the file to stable storage. Typically,
+// this means flushing the file system's in-memory copy of recently written data
+// to disk.
+func (r *Rotator) Sync() error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	if r.file == nil {
+		return nil
 	}
-	if err != nil {
-		return errors.Wrap(err, "failed to rotate backups")
+	return r.file.Sync()
+}
+
+// Rotate triggers a file rotation.
+func (r *Rotator) Rotate() error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	return r.rotate(rotateReasonManualTrigger)
+}
+
+// Close closes the currently open file.
+func (r *Rotator) Close() error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	return r.closeFile()
+}
+
+func (r *Rotator) dir() string {
+	return filepath.Dir(r.rot.ActiveFile())
+}
+
+func (r *Rotator) dirMode() os.FileMode {
+	mode := 0700
+	if r.permissions&0070 > 0 {
+		mode |= 0050
 	}
+	if r.permissions&0007 > 0 {
+		mode |= 0005
+	}
+	return os.FileMode(mode)
+}
+
+func (r *Rotator) closeFile() error {
+	if r.file == nil {
+		return nil
+	}
+	err := r.file.Close()
+	r.file = nil
+	return err
+}
 
-	return r.purgeOldBackups()
+type countRotator struct {
+	log             Logger
+	filename        string
+	intervalRotator *intervalRotator
+	maxBackups      uint
+}
+
+type dateRotator struct {
+	log             Logger
+	format          string
+	filenamePrefix  string
+	currentFilename string
+	intervalRotator *intervalRotator
 }
 
-func (r *Rotator) rotateByInterval(reason rotateReason) error {
-	fi, err := os.Stat(r.filename)
-	if os.IsNotExist(err) {
+func newRotater(log Logger, s SuffixType, filename string, maxBackups uint, interval time.Duration) rotater {
+	switch s {
+	case SuffixCount:
+		if interval > 0 {
+			return newIntervalRotator(log, interval, filename)
+		}
+		return &countRotator{
+			log:        log,
+			filename:   filename,
+			maxBackups: maxBackups,
+		}
+	case SuffixDate:
+		return &dateRotator{
+			log:             log,
+			format:          "2006010215:04:05.9999",
+			filenamePrefix:  filename,
+			currentFilename: filename + "TODO",
+		}
+	default:
```

Comment:
should we end up with nil as default? shouldnt the default be Count?

Reply:
In config it is already set to `SuffixCount`, but for extra safety we could add it here as well.

Reply:
yes i wanted to avoid having nil rotator and code panicking.